python/graphstorm/trainer/lp_trainer.py

```diff
@@ -220,27 +220,28 @@ def fit(self, train_loader, num_epochs,
                                  epoch, i, loss.item(), time.time() - batch_tic)
 
                 val_score = None
-                if self.evaluator is not None and \
-                    self.evaluator.do_eval(total_steps, epoch_end=False):
+                if self.can_do_model_eval(val_loader) and self.evaluator.do_eval(total_steps):
                     val_score = self.eval(model.module if is_distributed() else model,
                                           data, val_loader, test_loader, total_steps,
                                           edge_mask_for_gnn_embeddings, use_mini_batch_infer)
                     if self.evaluator.do_early_stop(val_score):
                         early_stop = True
 
-                # Every n iterations, check to save the top k models. If has validation score,
-                # will save the best top k. But if no validation, will either save
-                # the last k model or all models depends on the setting of top k
+                # In every save_model_frequency iterations, check to save the top k models.
+                # If has validation score, will save the best top k. If no validation, will
+                # either save the last k model or all models depends on the setting of top k.
                 if save_model_frequency > 0 and \
                     total_steps % save_model_frequency == 0 and \
                     total_steps != 0:
-                    if self.evaluator is None or val_score is not None:
-                        # We will save the best model when
-                        # 1. There is no evaluation, we will keep the
-                        #    latest K models.
-                        # 2. There is evaluaiton, we need to follow the
-                        #    guidance of validation score.
-                        self.save_topk_models(model, epoch, i, val_score, save_model_path)
+                    if self.can_do_model_eval(val_loader):
+                        # for model saving, force to do evaluation if can
+                        val_score = self.eval(model.module if is_distributed() else model,
```

Comment:
if eval_frequency == save_model_frequency, are we going to do the evaluation twice in the same iteration?